### ISSUE TITLE
Map blank nodes for multiple objects

### ISSUE THREAD

**Issue by simeonackermann** [semantifyit/RocketRML#17]
I'm failing to map multiple blank nodes as objects in multiple subjetcs...

Input:
```json
{
    "persons": [
        {
            "firstname": "Monika",
            "lastname": "Here",
            "speaks": [ "de", "en" ]
        },
        {
            "firstname": "John",
            "lastname": "Deere",
            "speaks": [ "en" ]
        },
        {
            "firstname": "Ruben",
            "lastname": "Bartoli",
            "speaks": [ "de", "fr", "ru" ]
        }
    ]
}
```

Mapping:
```yaml
prefixes:
  foaf: http://xmlns.com/foaf/0.1/
  ex: http://example.com/

mappings:
  person:
    sources:
      - ['input~jsonpath', '$.persons[*]']
    s: ex:$(firstname)_$(lastname)
    po:
      - [a, foaf:Person]
      - p: ex:speakes
        o:
          - mapping: languages

  languages:
    sources:
      - ['input~jsonpath', '$.persons[*].speaks[*]']
    s: null
    po:
      - [a, ex:Language]
      - [rdfs:label, $(.)]
```

Now the result is, that every person speaks every language ...  How can solve this?

**Comment by ThibaultGerrier** [comment]
Hello,
If you want to join 2 mappings (your case) it will by default join everything together (cartesian product). To limit that we use join-conditions:

In turtle rml there is rr:joinCondition to accomplish that.

In yarrrml it is a bit more verbose but you do this via a condition: (some example: https://rml.io/yarrrml/spec/#referring-to-other-mappings)

So in your case, your mapping with a condition could look like this:

```yaml
prefixes:
  foaf: http://xmlns.com/foaf/0.1/
  ex: http://example.com/

mappings:
  person:
    sources:
      - ['input~jsonpath', '$.persons[*]']
    s: ex:$(firstname)_$(lastname)
    po:
      - [a, foaf:Person]
      - p: ex:speakes
        o:
          - mapping: languages
            condition:
              function: equal
              parameters:
                - [str1, $(firstname), s]
                - [str2, $(^^firstname), o]

  languages:
    sources:
      - ['input~jsonpath', '$.persons[*].speaks[*]']
    s: null
    po:
      - [a, ex:Language]
      - [rdfs:label, $(.)]

```

Join conditions are a bit tricky to understand. Two mappings work completely independent from each other.

You define a json path from the current subject (the parameter with the "s" - "firstname") and a second json path from the 2nd mapping (the parameter with the "o" - "^^firstname").
The code will then loop over all entities of the first mapping and all of the 2nd mapping and then only add the links, wherever the results of the 2 json paths are the same.


To show what will happen:

First the person mapping will run and iterate on all `$.persons[*]`
- create Person for `$.persons.0`
- create Person for `$.persons.1`
- create Person for `$.persons.1`

Next the language mapping (with iterator `$.persons[*].speaks[*]`)
- create Langauge for `$.persons.0.speaks.0`
- create Langauge for `$.persons.0.speaks.1`
- create Langauge for `$.persons.1.speaks.0`
- create Langauge for `$.persons.2.speaks.0`
- create Langauge for `$.persons.2.speaks.1`
- create Langauge for `$.persons.2.speaks.2`

Next add ex:speakes wherever the join condition holds true:

loop over all possible combinations between persons and languages (cartesian product)

take the first joincondition parameter with each person and the second joincondition parameter for each language:

(note, the `^`character in the second json path means "go back one level", it's not official jsonpath, but is supported by the library we use to execute json paths (jsonpath plus). Other implementations of rml might not support this.
We use 2 `^` since we need to go "up" 2 levels, from `$.persons[*].speaks[*]` to `$.persons[*].speaks` (the array) to `$.persons[*]` (the object))

- `$.persons.0.firstname` == `$.persons.0.speaks.0.^^firstname` -> add ex:speak between the two 
- `$.persons.0.firstname` == `$.persons.0.speaks.1.^^firstname` -> add ex:speak between the two
- `$.persons.0.firstname` != `$.persons.1.speaks.0.^^firstname` -> skip
- `$.persons.0.firstname` != `$.persons.2.speaks.0.^^firstname` -> skip
...
- `$.persons.2.firstname` != `$.persons.2.speaks.2` -> add


Note that we had to use "firstname" in the join condition since we needed a common jsonpath value to join on. If the values of "firstname" is not unique, we would have a problem. Usually you would use unique ids to join.
For cases with no unique fiels (such as your example), we added a feature, which lets you returns the path to the current element: "PATH~". (since jsonpath natively doesn't have such a feature). You can add a path after the "~" to change the current json path element, such as go into a child or a parent.

You can use it such as: `- [ex:path, $(PATH~)]`, which would return the JSON-pointer to the current element: e.g. "/persons/0"

Usage within a joincondition:
```yaml
      - p: ex:speakes
        o:
          - mapping: languages
            condition:
              function: equal
              parameters:
                - [str1, $(PATH~), s]
                - [str2, $(PATH~^^), o]
```
basically means: whenever the path of the current person element, and the path of the parent of the parent of the current language element are the same, then join.
(Note again, that this only works for RocketRML and is not RML nor JSON-path standard)

I hope this helped a bit:)

**Comment by simeonackermann** [comment]
Wow, thanks for the extensive answer! That really helps a lot and fixed my problem for now, although it requires jsonpath-plus and RocketRML specific `PATH~` implementation.
